src/libcollections/string.rs

```diff
@@ -1881,6 +1881,41 @@ impl Into<Vec<u8>> for String {
     }
 }
 
+#[stable(feature = "stringfromchars", since = "1.12.0")]
+impl<'a> From<&'a [char]> for String {
+    #[inline]
+    fn from(v: &'a [char]) -> String {
+        let mut s = String::with_capacity(v.len());
+        for c in v {
+            s.push(*c);
+        }
+        s
+    }
+}
+
+#[stable(feature = "stringfromchars", since = "1.12.0")]
+impl From<Vec<char>> for String {
+    fn from(mut v: Vec<char>) -> String {
+        unsafe {
+            let cap = v.capacity();
+            let ptr = v.as_mut_ptr() as *mut u8;
+            let mut bytes = 0usize;
+
+            for chr in v.iter() {
+                // Regular loop instead of copy_nonoverlapping, because LLVM insists on having a
+                // memcpy for slices shorter than 4 bytes.
+                for b in chr.encode_utf8().as_slice() {
+                    // Neither bytes nor ptr can overflow.
+                    *ptr.offset(bytes as isize) = *b;
+                    bytes = bytes.wrapping_add(1);
```

Comment:
Right, I misread what `bytes` was used for.

Why does debug mode matter? IMO `+` is the operator to use if you know it can't overflow, the wrapping ops are for when you know it can overflow and that's what you want.